### ISSUE TITLE
Sphinx and Napoleon do not link property types

### ISSUE THREAD

**Issue by AWhetter** [sphinx-doc/sphinx#2979]
There's no official way of documenting the type of a property currently. Using `:var:` and `:vartype:` is for attributes only. `:type:` is for parameters although I've seen it used to document property types also in the past.

The example for Google Style docstrings shows that the type of a property is documented with `"""type: Summary."""`. This does not get linked properly with intersphinx though and the type isn't rendered in it's own field.

As a side issue, for attribute Napoleon itself currently sets the type by inserting the type in italics. So they also aren't linked. Perhaps it could use the same field that is decided upon for properties.

**Comment by RobRuana** [comment]
Distinguishing a property from an attribute in documentation has definitely been an issue. I don't see a huge problem with using `:var:` and `:vartype:` for properties, because from an external perspective they have the same usage.

An attribute could be refactored into a property, and the docs wouldn't have to change, right? I may be missing something.

**Comment by AWhetter** [comment]
An attribute could definitely be refactored into a property. In fact I think it is weird to have them shown differently in the documentation when they have the same usage, as you say.

If we did that and used `:vartype:` for property types then I think you're right the the docs wouldn't need to change. I say `:vartype:` only because I would prefer to keep the docstrings for properties inside the property getter, and not force the usage of `:var:` and `:vartype:` on the class docstring.

**Comment by RobRuana** [comment]
Napoleon supports documenting attributes in the class docstring like so:

```
    Attributes:
        attr1 (str): Description of `attr1`.
        attr2 (int): Description of `attr2`.
```

or on the property directly like so:

```
    @property
    def readonly_property(self):
        """str: Properties should be documented in their getter method."""
        return 'readonly_property'

    @property
    def readwrite_property(self):
        """list(str): Properties with both a getter and setter
        should only be documented in their getter method.

        If the setter method contains notable behavior, it should be
        mentioned here.
        """
        return ['readwrite_property']
```

Is this what you are looking for? Or do you have any suggestions for changes or improvements to the current style?

**Comment by AWhetter** [comment]
I think that style is great.
It's maybe a little annoying to have the property type take up space in the summary line but I think it would be worse to need a separate header just to define the type. So I say keep it as is.

**Comment by RobRuana** [comment]
Ok, sorry for the delay! I'm not entirely clear what this issue is asking for. Can you describe specifically what the issue to be resolved is?

**Comment by AWhetter** [comment]
For napoleon docstrings, using the syntax in the examples (`"""type: Summary"""`) renders the docstring as it appears. The type doesn't get put into it's own field and it doesn't get linked automatically by intersphinx either.

For rst docstrings, there's no way to mark the type of a property and have it link automatically. `type` is often used but isn't an official field for properties so intersphinx doesn't link the type. `rtype` links but having a "Return type" for something that it both get and set doesn't really make sense.

`var`/`vartype` and properties being rendered differently is a separate issue really.

**Comment by RobRuana** [comment]
Can you post some example code demonstrating the behavior you're interested in changing? Also, can you post your conf.py settings?

**Comment by AWhetter** [comment]
I've uploaded an example here: https://files.awhetter.co.uk/sphinx_issue_2979_example.tar.gz

It demonstrates how variables and properties are rendered with sphinx and Napoleon, and comments on what ideally would be different about them. Let me know if they don't make sense.

**Comment by RobRuana** [comment]
Okay, I think I fully understand this issue now. Your comments make perfect sense. I initially thought Napoleon was doing something wrong, but this issue must be fixed at a deeper level, probably in autodoc.

As you stated, the real heart of the issue is that there's no way to define a  type for a `@property` using reStructuredText. I'd be okay with something like the following:

```
class PropertyTypeProposals(object):

    @property
    def prop1(self):
        """This uses the :type: field to specify a type
        :type: str
        """
        return 'prop1'
```

I don't have the time to implement something like the above, so I'm going to unassign myself from this issue. But once property types are supported, I can update napoloeon to produce the correct reStructuredText markup.

# In the meantime...

As a workaround, you can still document properties in the `Attributes` section of the class level docstring.

In the sample project you provided there were a couple issues that were preventing intersphinx from working:

```
class Test(object):
    """This demonstrates how sphinx renders types.

    Variables:
        var1 (str): This is not linked, and rendered differently to a property.
    """
```

The `Variables` section should be labelled `Attributes` instead. Also, to get intersphinx to recognize your attribute list, you need to set `napoleon_use_ivar` to `True` in your `conf.py`.

You should now have proper hyperlinking.

**Comment by AWhetter** [comment]
I can't come up with a nice implementation of this.

The problem is that both for autodoc and napoleon, they can't add to the variables section easily because it would mean that the properties change the docstring of the class. So really this will only ever work if properties are documented with `.. attribute::`. So `napoleon_use_ivar` would have to be False.

#4045 implements linking attributes with the following syntax `:class:str -- This is an rst property`, which looks great. The problem is that in autodoc you have to write this manually in the docstring. autodoc can't know the type of the property without having extract a field like `:type:` out of the docstring, which seems hacky.

#4047 instead does the implementation of using `:type:` to document properties. This is a little hacky because, as I explain on the pull request, it doesn't define `:type:` as it's own directive because it already exists. I also don't like how this looks as much, but I don't think that there's a way of making it render like #4045 because it would mean that a directive would affect a line earlier in the docstring. I've had to edit some of the parsing code that isn't domain specific so we can't really mess with the docstring in this way.

Overall I think #4047 gives us the best we can get without a load of refactoring work. It's good enough for me!

**Comment by AWhetter** [comment]
Is there a preference on either of the above methods? Or is there an even better solution that I'm missing?

**Comment by AWhetter** [comment]
Closed via #4047
